cpp/src/arrow/compute/exec/hash_join_node_test.cc

```diff
@@ -1113,5 +1113,539 @@ TEST(HashJoin, Random) {
+      auto prefix_batch,
+      ExecBatch::Make({expected_batch.values[0].array()->Slice(
+                           0, expected_batch.length - expected_num_r_no_match),
+                       expected_batch.values[1].array()->Slice(
+                           0, expected_batch.length - expected_num_r_no_match),
+                       expected_batch.values[2].array()->Slice(
+                           0, expected_batch.length - expected_num_r_no_match),
+                       expected_batch.values[3].array()->Slice(
+                           0, expected_batch.length - expected_num_r_no_match)}));
+  for (int i = 0; i < (parallel ? batch_multiplicity_for_parallel : 1); ++i) {
+    expected_batches.push_back(prefix_batch);
+  }
+  if (expected_num_r_no_match > 0) {
+    ASSERT_OK_AND_ASSIGN(
+        auto suffix_batch,
+        ExecBatch::Make({expected_batch.values[0].array()->Slice(
+                             expected_batch.length - expected_num_r_no_match,
+                             expected_num_r_no_match),
+                         expected_batch.values[1].array()->Slice(
+                             expected_batch.length - expected_num_r_no_match,
+                             expected_num_r_no_match),
+                         expected_batch.values[2].array()->Slice(
+                             expected_batch.length - expected_num_r_no_match,
+                             expected_num_r_no_match),
+                         expected_batch.values[3].array()->Slice(
+                             expected_batch.length - expected_num_r_no_match,
+                             expected_num_r_no_match)}));
+    expected_batches.push_back(suffix_batch);
+  }
+
+  ASSERT_OK_AND_ASSIGN(std::shared_ptr<Table> expected,
+                       TableFromExecBatches(output_schema, expected_batches));
+
+  // Compare results
+  AssertTablesEqual(expected, output);
+
+  // TODO: This was added for debugging. Remove in the final version.
+  // std::cout << output->ToString();
+}
+
+TEST(HashJoin, Dictionary) {
+  auto int8_utf8 = std::make_shared<DictionaryType>(int8(), utf8());
+  auto uint8_utf8 = std::make_shared<DictionaryType>(uint8(), utf8());
+  auto int16_utf8 = std::make_shared<DictionaryType>(int16(), utf8());
+  auto uint16_utf8 = std::make_shared<DictionaryType>(uint16(), utf8());
+  auto int32_utf8 = std::make_shared<DictionaryType>(int32(), utf8());
+  auto uint32_utf8 = std::make_shared<DictionaryType>(uint32(), utf8());
+  auto int64_utf8 = std::make_shared<DictionaryType>(int64(), utf8());
+  auto uint64_utf8 = std::make_shared<DictionaryType>(uint64(), utf8());
+  std::shared_ptr<DictionaryType> dict_types[] = {int8_utf8,   uint8_utf8, int16_utf8,
+                                                  uint16_utf8, int32_utf8, uint32_utf8,
+                                                  int64_utf8,  uint64_utf8};
+
+  Random64Bit rng(43);
+
+  // Dictionaries in payload columns
+  for (auto parallel : {false, true})
```

Comment:
nit: can we insert braces here for readability?

Reply:
added